sdks/python/apache_beam/runners/dataflow/internal/dependency.py

```diff
@@ -453,36 +458,81 @@ def _build_setup_package(setup_file, temp_dir, build_setup_args=None):
+      Set to 'pypi' to download and stage a wheel and source SDK from PyPi.
+    staging_location: A GCS bucket where the SDK file should be copied.
     temp_dir: path to temporary location where the file should be downloaded.
 
+  Returns:
+    A list of SDK files that were staged to the staging location.
+
   Raises:
-    RuntimeError: If wget on the URL specified returs errors or the file
-      cannot be copied from/to GCS.
+    RuntimeError: if staging was not successful.
   """
   if (sdk_remote_location.startswith('http://') or
       sdk_remote_location.startswith('https://')):
-    logging.info(
-        'Staging Beam SDK tarball from %s to %s',
-        sdk_remote_location, staged_path)
     local_download_file = _dependency_file_download(
         sdk_remote_location, temp_dir)
+    staged_name = _desired_sdk_filename_in_staging_location(local_download_file)
+    staged_path = FileSystems.join(staging_location, staged_name)
+    logging.info(
+        'Staging Beam SDK from %s to %s',
+        sdk_remote_location, staged_path)
     _dependency_file_copy(local_download_file, staged_path)
+    return [staged_name]
   elif sdk_remote_location.startswith('gs://'):
     # Stage the file to the GCS staging area.
+    staged_name = _desired_sdk_filename_in_staging_location(sdk_remote_location)
+    staged_path = FileSystems.join(staging_location, staged_name)
     logging.info(
-        'Staging Beam SDK tarball from %s to %s',
+        'Staging Beam SDK from %s to %s',
         sdk_remote_location, staged_path)
     _dependency_file_copy(sdk_remote_location, staged_path)
+    return [staged_name]
   elif sdk_remote_location == 'pypi':
-    logging.info('Staging the SDK tarball from PyPI to %s', staged_path)
-    _dependency_file_copy(_download_pypi_sdk_package(temp_dir), staged_path)
+    sdk_local_file = _download_pypi_sdk_package(temp_dir)
+    sdk_sources_staged_name = _desired_sdk_filename_in_staging_location(
+        sdk_local_file)
+    staged_path = FileSystems.join(staging_location, sdk_sources_staged_name)
+    logging.info('Staging SDK sources from PyPI to %s', staged_path)
+    _dependency_file_copy(sdk_local_file, staged_path)
+    staged_sdk_files = [sdk_sources_staged_name]
+    try:
+      # Stage binary distribution of the SDK, for now on a best-effort basis.
+      sdk_local_file = _download_pypi_sdk_package(temp_dir, fetch_binary=True)
+      sdk_binary_staged_name = _desired_sdk_filename_in_staging_location(
+          sdk_local_file)
+      staged_path = FileSystems.join(staging_location, sdk_binary_staged_name)
+      logging.info('Staging binary distribution of the SDK from PyPI to %s',
+                   staged_path)
+      _dependency_file_copy(sdk_local_file, staged_path)
+      staged_sdk_files.append(sdk_binary_staged_name)
+    except RuntimeError as e:
+      logging.warn('Failed to download requested binary distribution '
+                   'of the SDK: %s', repr(e))
```

Comment:
For some reason repr(e) does not keep the exception message that CalledProcessError contains:
This is what it looks like: WARNING:root:Failed to download requested binary distribution of the SDK: RuntimeError('CalledProcessError()',). There is a command line in the logs above, so it's clear what goes wrong, but I was wondering what is a better way of passing exception message here?

Reply:
`traceback.format_exc()` perhaps? Please do a follow up update or open a JIRA.